### ISSUE TITLE
Windows Multithreaded environment vc12 build configuration

### ISSUE THREAD

**Issue by arkhivania** [Itseez/opencv#5882]
Hello! I have not very nice situation with my little test:
There is a metacode from c# project
```csharp
[Test]
        [TestCase(1024, 1024, 256, 256, NativePixelFormat.Y, Interpolation.Nearest)]
        [TestCase(1024, 1024, 256, 256, NativePixelFormat.Y, Interpolation.Linear)]
        [TestCase(1024, 1024, 256, 256, NativePixelFormat.Y2, Interpolation.Nearest)]
        [TestCase(1024, 1024, 256, 256, NativePixelFormat.Y2, Interpolation.Linear)]
        public void Resize(int sourceWidth, int sourceHeight, int destWidth, int destHeight, NativePixelFormat format, Interpolation interpolation)
        {
            Parallel.For(0, 1000, i => {
                using (var source = new ImageBuffer(new StoreInfo { Width = sourceWidth, Height = sourceHeight, Format = format }, new Framework.HeapMemoryPool()))
                using (var dest = new ImageBuffer(new StoreInfo { Width = destWidth, Height = destHeight, Format = format }, new Framework.HeapMemoryPool()))
                    ImageMath.Resize.Process(source.ToNativeImageBuffer(), dest.ToNativeImageBuffer(), interpolation);
            });
        }
```

Resize.Process - it's PInvoke to cv::resize method

Unfortunately, there are access violation memory exceptions fired. The situation appears only with parallel execution. 

cv::warpAffine have the same trouble. I didn't compile opencv myself, it's vc12 build from OpenCV 3.0 public distr. I examined some code, and have only one idea. WarpAffine use AutoBuffer<int> in function, maby some strange compiler optimization crash the stack in multithreaded environment.

**Comment by alalek** [comment]
More details are definitely required: configuration dump (see `cv::getBuildInformation()`), stack traces (use Debug build).

It will be great to have reproducer with C++ code only (you may use `std::thread` to minimize code) to eliminate possible errors in C# bindings.

**Comment by arkhivania** [comment]
[out.txt](https://github.com/Itseez/opencv/files/72212/out.txt)

![stack](https://cloud.githubusercontent.com/assets/5111153/12003645/34c2faa6-ab39-11e5-8321-25ffaed5222b.png)

Is there something wrong with OpenCL initialization in diffrent threads?

[OcvTroubles.zip](https://github.com/Itseez/opencv/files/72215/OcvTroubles.zip)

**Comment by arkhivania** [comment]
Ok! Workaround it's initialize opencl singletons and other stuff in main application thread on startup.
```C++
#include "stdafx.h"

#include "opencv2/imgproc/imgproc.hpp"
#include <ppl.h>
#include <ostream>

using namespace concurrency;

void TestImageResize()
{
	cv::Mat sm1(1024, 1024, CV_8U);
	cv::Mat sm2(1024, 1024, CV_8U);

	cv::Mat tm1(128, 128, CV_8U);
	cv::Mat tm2(128, 128, CV_8U);

	cv::resize(sm1, tm1, tm1.size());
	cv::resize(sm2, tm2, tm2.size());
}

int _tmain(int argc, _TCHAR* argv[])
{
	auto bc = cv::getBuildInformation();
	printf(bc.c_str());

	TestImageResize();

	parallel_for(0, 10000, [&](int index)
	{
		TestImageResize();
	});

	return 0;
}
````

**Comment by alalek** [comment]
I'm able to reproduce this issue with 3.0.0.
Also this line helps to workaround this issue:
```
auto useOpenCL = cv::ocl::useOpenCL();
```

This bug should be fixed in 3.1.0 (PR #4150)

**Comment by arkhivania** [comment]
Thanks!
